File: cookie_finder/gimbal/pan_tilt.py
"""
High-level pan/tilt gimbal control using two stepper motors.

This module provides a unified interface for controlling a pan/tilt camera gimbal
with two 28BYJ-48 stepper motors (pan and tilt), each with Center-Off SPDT
limit switch detection (min + max channels, active-low).

GPIO Pin Layout (Orange Pi Zero 2W, gpiochip1 offsets):
  Pan Motor IN1-IN4: 271, 268, 258, 272 (physical 31/33/35/37)
  Pan Limit min/max: 257 / 227 (physical 12 / 13)
  Tilt Motor IN1-IN4: 262, 229, 233, 265 (physical 22/24/26/28)
  Tilt Limit min/max: 261 / 270 (physical 15 / 16)
  UART0 reserved: physical 8 / 10
"""

import logging
import threading
from cookie_finder.gimbal.config import load_gimbal_config
from cookie_finder.gimbal.stepper import StepperMotor, MotorDirection

logger = logging.getLogger(__name__)

if not logging.root.handlers:
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s %(levelname)s [%(name)s] %(message)s",
    )


class PanTiltGimbal:
    """
    Unified pan/tilt gimbal control.

    Manages two stepper motors for pan (horizontal) and tilt (vertical) rotation.
    Provides synchronized movement, position tracking, and limit switch safety.
    Thread-safe via internal locks.
    """

    # GPIO pin assignments for Orange Pi Zero 2W (gpiochip1 offsets)
    PAN_CONTROL_PINS = (271, 268, 258, 272)   # IN1-IN4 (PI15, PI12, PI02, PI16)
    PAN_LIMIT_MIN_PIN = 257                   # physical 12, PI01 (CCW / home)
    PAN_LIMIT_MAX_PIN = 227                   # physical 13, RXD.5 (CW)

    TILT_CONTROL_PINS = (262, 229, 233, 265)  # IN1-IN4 (RXD.2, CE.0, CE.1, SCL.2)
    TILT_LIMIT_MIN_PIN = 261                  # physical 15, TXD.2 (CCW / home)
    TILT_LIMIT_MAX_PIN = 270                  # physical 16, PWM4 (CW)

    def __init__(self, max_pan: float = 180.0, max_tilt: float = 180.0):
        """
        Initialize pan/tilt gimbal.

        Args:
            max_pan: Maximum pan angle in degrees (0-180)
            max_tilt: Maximum tilt angle in degrees (0-180)
        """
        self.max_pan = max_pan
        self.max_tilt = max_tilt
        logger.info("PanTiltGimbal.__init__(max_pan=%.1f, max_tilt=%.1f)", max_pan, max_tilt)

        gimbal_config = load_gimbal_config()
        pan_phase = tuple(gimbal_config["pan_phase_order"])
        tilt_phase = tuple(gimbal_config["tilt_phase_order"])

        # Create motor controllers
        self.pan_motor = StepperMotor(
            control_pins=self.PAN_CONTROL_PINS,
            limit_min_pin=self.PAN_LIMIT_MIN_PIN,
            limit_max_pin=self.PAN_LIMIT_MAX_PIN,
            max_angle=max_pan,
            motor_name="Pan",
            phase_order=pan_phase,
        )

        self.tilt_motor = StepperMotor(
            control_pins=self.TILT_CONTROL_PINS,
            limit_min_pin=self.TILT_LIMIT_MIN_PIN,
            limit_max_pin=self.TILT_LIMIT_MAX_PIN,
            max_angle=max_tilt,
            motor_name="Tilt",
            phase_order=tilt_phase,
        )

        # State tracking
        self._lock = threading.Lock()
        self._is_calibrated = False

    def set_speed(self, pan_hz: float = 500, tilt_hz: float = 500) -> None:
        """
        Set stepping frequency for both motors.

        Args:
            pan_hz: Pan motor stepping frequency (500-2000 Hz recommended)
            tilt_hz: Tilt motor stepping frequency
        """
        logger.info("set_speed(pan_hz=%.1f, tilt_hz=%.1f)", pan_hz, tilt_hz)
        with self._lock:
            self.pan_motor.set_speed(pan_hz)
            self.tilt_motor.set_speed(tilt_hz)

    def move_to_angles(self, pan_angle: float, tilt_angle: float) -> None:
        """
        Move gimbal to specified pan and tilt angles.

        Arms are moved in parallel and independently.

        Args:
            pan_angle: Target pan angle in degrees (0 to max_pan)
            tilt_angle: Target tilt angle in degrees (0 to max_tilt)
        """
        logger.info("move_to_angles(pan=%.1f, tilt=%.1f)", pan_angle, tilt_angle)
        with self._lock:
            pan_angle = max(0, min(pan_angle, self.max_pan))
            tilt_angle = max(0, min(tilt_angle, self.max_tilt))

            self.pan_motor.move_to_angle(pan_angle)
            self.tilt_motor.move_to_angle(tilt_angle)

    def pan(self, angle: float) -> None:
        """
        Move pan motor to angle, leave tilt unchanged.

        Args:
            angle: Pan angle in degrees
        """
        logger.info("pan(angle=%.1f)", angle)
        with self._lock:
            self.pan_motor.move_to_angle(angle)

    def tilt(self, angle: float) -> None:
        """
        Move tilt motor to angle, leave pan unchanged.

        Args:
            angle: Tilt angle in degrees
        """
        logger.info("tilt(angle=%.1f)", angle)
        with self._lock:
            self.tilt_motor.move_to_angle(angle)

    def pan_step(self, direction: int, steps: int = 1) -> None:
        """
        Step pan motor incrementally.

        Args:
            direction: 1 for clockwise, -1 for counterclockwise
            steps: Number of half-steps
        """
        logger.info("pan_step(direction=%d, steps=%d)", direction, steps)
        with self._lock:
            motor_dir = MotorDirection.CLOCKWISE if direction > 0 else MotorDirection.COUNTERCLOCKWISE
            self.pan_motor.step(motor_dir, steps)

    def tilt_step(self, direction: int, steps: int = 1) -> None:
        """
        Step tilt motor incrementally.

        Args:
            direction: 1 for up, -1 for down
            steps: Number of half-steps
        """
        logger.info("tilt_step(direction=%d, steps=%d)", direction, steps)
        with self._lock:
            motor_dir = MotorDirection.CLOCKWISE if direction > 0 else MotorDirection.COUNTERCLOCKWISE
            self.tilt_motor.step(motor_dir, steps)

    def get_position(self) -> tuple:
        """
        Get current pan and tilt angles.

        Returns:
            (pan_angle, tilt_angle) in degrees
        """
        with self._lock:
            position = (self.pan_motor.get_angle(), self.tilt_motor.get_angle())
        logger.debug("get_position() -> (%.1f, %.1f)", position[0], position[1])
        return position

    def is_moving(self) -> bool:
        """Check if either motor is currently moving."""
        with self._lock:
            moving = self.pan_motor.is_moving or self.tilt_motor.is_moving
        logger.debug("is_moving() -> %s", moving)
        return moving

    def home(self) -> None:
        """
        Calibrate gimbal by moving both motors to home position (0, 0).
        Uses min limit switches for accurate calibration.

        Blocks until both motors are homed.
        """
        logger.info("home() starting")
        with self._lock:
            self._is_calibrated = False

        # Home pan first
        self.pan_motor.home()

        # Home tilt
        self.tilt_motor.home()

        with self._lock:
            self._is_calibrated = True

        logger.info("home() complete (0°, 0°)")

    def is_calibrated(self) -> bool:
        """Check if gimbal has been calibrated."""
        with self._lock:
            calibrated = self._is_calibrated
        logger.debug("is_calibrated() -> %s", calibrated)
        return calibrated

    def stop(self) -> None:
        """Stop both motors and hold current position."""
        logger.info("stop()")
        with self._lock:
            self.pan_motor.stop()
            self.tilt_motor.stop()

    def cleanup(self) -> None:
        """Clean up GPIO resources and stop motors."""
        logger.info("cleanup()")
        with self._lock:
            self.pan_motor.cleanup()
            self.tilt_motor.cleanup()
        logger.info("cleanup() complete")

File: cookie_finder_rust/cookie-finder-ctl/src/gimbal/pan_tilt.rs
use super::stepper::{DriveMode, StepperMotor};
use crate::config::{
    validate_phase_order, DEFAULT_PAN_HZ, DEFAULT_TILT_HZ, GimbalConfig, MAX_PAN, MAX_TILT,
    PAN_CONTROL_PINS, PAN_LIMIT_MAX_PIN, PAN_LIMIT_MIN_PIN, TILT_CONTROL_PINS, TILT_LIMIT_MAX_PIN,
    TILT_LIMIT_MIN_PIN,
};
use std::sync::Mutex;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MotorId {
    Pan,
    Tilt,
}

pub struct PanTiltGimbal {
    pub max_pan: f64,
    pub max_tilt: f64,
    inner: Mutex<Inner>,
}

struct Inner {
    pan: StepperMotor,
    tilt: StepperMotor,
    commanded_pan: f64,
    commanded_tilt: f64,
}

impl PanTiltGimbal {
    pub fn new(config: &GimbalConfig) -> Self {
        Self {
            max_pan: MAX_PAN,
            max_tilt: MAX_TILT,
            inner: Mutex::new(Inner {
                pan: StepperMotor::new(
                    "Pan",
                    PAN_CONTROL_PINS,
                    config.pan_phase_order,
                    MAX_PAN,
                    Some((PAN_LIMIT_MIN_PIN, PAN_LIMIT_MAX_PIN)),
                ),
                tilt: StepperMotor::new(
                    "Tilt",
                    TILT_CONTROL_PINS,
                    config.tilt_phase_order,
                    MAX_TILT,
                    Some((TILT_LIMIT_MIN_PIN, TILT_LIMIT_MAX_PIN)),
                ),
                commanded_pan: 0.0,
                commanded_tilt: 0.0,
            }),
        }
    }

    pub fn set_speed(&self, pan_hz: f64, tilt_hz: f64) {
        let mut g = self.inner.lock().unwrap();
        g.pan.set_speed(pan_hz);
        g.tilt.set_speed(tilt_hz);
    }

    pub fn move_to_angles(&self, pan: f64, tilt: f64) {
        let pan = pan.clamp(0.0, self.max_pan);
        let tilt = tilt.clamp(0.0, self.max_tilt);
        let mut g = self.inner.lock().unwrap();
        g.commanded_pan = pan;
        g.commanded_tilt = tilt;
        g.pan.set_target(pan);
        g.tilt.set_target(tilt);
    }

    pub fn pan_step(&self, direction: i32, steps: u32) {
        let mut g = self.inner.lock().unwrap();
        g.pan.step_fixed(direction, steps);
        g.commanded_pan = g.pan.get_angle();
    }

    pub fn tilt_step(&self, direction: i32, steps: u32) {
        let mut g = self.inner.lock().unwrap();
        g.tilt.step_fixed(direction, steps);
        g.commanded_tilt = g.tilt.get_angle();
    }

    pub fn get_position(&self) -> (f64, f64) {
        let g = self.inner.lock().unwrap();
        (g.pan.get_angle(), g.tilt.get_angle())
    }

    pub fn get_commanded(&self) -> (f64, f64) {
        let g = self.inner.lock().unwrap();
        (g.commanded_pan, g.commanded_tilt)
    }

    pub fn get_phase_orders(&self) -> ([usize; 4], [usize; 4]) {
        let g = self.inner.lock().unwrap();
        (g.pan.phase_order(), g.tilt.phase_order())
    }

    pub fn set_phase_order(&self, motor: MotorId, order: [usize; 4]) -> anyhow::Result<()> {
        let name = match motor {
            MotorId::Pan => "pan",
            MotorId::Tilt => "tilt",
        };
        validate_phase_order(&order, &format!("{name}_phase_order"))?;
        let mut g = self.inner.lock().unwrap();
        match motor {
            MotorId::Pan => g.pan.set_phase_order(order),
            MotorId::Tilt => g.tilt.set_phase_order(order),
        }
        Ok(())
    }

    /// Drive mode is shared by pan and tilt (same energization algorithm).
    pub fn get_drive_mode(&self) -> DriveMode {
        self.inner.lock().unwrap().pan.drive_mode()
    }

    pub fn set_drive_mode(&self, mode: DriveMode) {
        let mut g = self.inner.lock().unwrap();
        g.pan.set_drive_mode(mode);
        g.tilt.set_drive_mode(mode);
    }

    pub fn is_moving(&self) -> bool {
        let g = self.inner.lock().unwrap();
        g.pan.is_moving() || g.tilt.is_moving()
    }

    pub fn home(&self) {
        let mut g = self.inner.lock().unwrap();
        g.pan.home();
        g.tilt.home();
        g.commanded_pan = 0.0;
        g.commanded_tilt = 0.0;
    }

    pub fn stop(&self) {
        let mut g = self.inner.lock().unwrap();
        g.pan.stop();
        g.tilt.stop();
        g.commanded_pan = g.pan.get_angle();
        g.commanded_tilt = g.tilt.get_angle();
    }

    pub fn disable_motors(&self) {
        let mut g = self.inner.lock().unwrap();
        g.pan.disable_coils();
        g.tilt.disable_coils();
        g.commanded_pan = g.pan.get_angle();
        g.commanded_tilt = g.tilt.get_angle();
    }

    pub fn tick(&self) {
        let mut g = self.inner.lock().unwrap();
        if g.pan.is_moving() {
            g.pan.tick();
        }
        if g.tilt.is_moving() {
            g.tilt.tick();
        }
    }

    pub fn pan_hz(&self) -> f64 {
        self.inner.lock().unwrap().pan.speed_hz()
    }

    pub fn tilt_hz(&self) -> f64 {
        self.inner.lock().unwrap().tilt.speed_hz()
    }

    pub fn cleanup(&self) {
        let mut g = self.inner.lock().unwrap();
        g.pan.cleanup();
        g.tilt.cleanup();
    }
}

impl Default for PanTiltGimbal {
    fn default() -> Self {
        let g = Self::new(&GimbalConfig::default());
        g.set_speed(DEFAULT_PAN_HZ, DEFAULT_TILT_HZ);
        g
    }
}
